tests/test_shardformer/test_layer/test_sequence_parallel.py

```diff
@@ -0,0 +1,243 @@
+            value (Tensor): value input to the layer
+            args: other args
+
+        Returns:
+            * output (Tensor): context output
+        """
+
+        # in shape : e.g.,  [s/p:h:]
+        query = self.q(x)
+        key = self.k(x)
+        value = self.v(x)
+        # TODO Merge three alltoall calls into one
+        query_layer = all_to_all_comm(query, self.spg, self.scatter_idx, self.gather_idx)
+        key_layer = all_to_all_comm(key, self.spg, self.scatter_idx, self.gather_idx)
+        value_layer = all_to_all_comm(value, self.spg, self.scatter_idx, self.gather_idx)
+
+        # out shape : e.g., [s:h/p:]
+        attn_score = self.attn(query_layer, key_layer, value_layer)
+
+        output = all_to_all_comm(attn_score, self.spg, self.gather_idx, self.scatter_idx)
+
+        # output e.g., [s/p::h]
+        output = self.out(output)
+
+        return output
+
+
+class MultiHeadAttn(nn.Module):
+    def __init__(self, head_num, hidden_dim, q_proj, k_proj, v_proj, out_proj):
+        super(MultiHeadAttn, self).__init__()
+        self.head_num = head_num
+        self.hidden_dim = hidden_dim
+        assert hidden_dim % head_num == 0
+        self.head_dim = hidden_dim // head_num
+
+        self.q = q_proj
+        self.k = k_proj
+        self.v = v_proj
+        self.out = out_proj
+
+    def attn(self, q, k, v):
+        batch_size, seq_len = q.shape[0], q.shape[1]
+
+        scale = self.head_dim**0.5
+        qk = torch.matmul(q, k.transpose(-2, -1)) / scale
+
+        # if attn_mask is not None:
+        #     mask = attn_mask == 0
+        #     qk[mask] = torch.tensor(float('-inf'))
+
+        weights = F.softmax(qk, dim=-1)
+
+        attention_score = torch.matmul(weights, v)
+
+        return attention_score
+
+    def split(self, x, batch_size, seq_len):
+        res = x.view(batch_size, seq_len, self.head_num, self.head_dim).transpose(1, 2)
+        return res
+
+    def forward(self, x):
+        batch_size, seq_len, hidden_dim = x.shape
+        assert hidden_dim == self.hidden_dim, "hidden_dim should be equal to self.hidden_dim"
+        query_mha = self.split(self.q(x), batch_size, seq_len)
+        key_mha = self.split(self.k(x), batch_size, seq_len)
+        value_mha = self.split(self.v(x), batch_size, seq_len)
+        score_mha = self.attn(query_mha, key_mha, value_mha)
+        score_mha_final = score_mha.transpose(1, 2).contiguous().view(batch_size, -1, self.hidden_dim)
+        output_mha = self.out(score_mha_final)
+
+        return output_mha
+
+
+def seq_parallel_attn(seq_len, hidden_dim, head_num, batch_size):
+    seq_len = seq_len
+    hidden_dim = hidden_dim
+    head_num = head_num
+    batch_size = batch_size
+    world_size = dist.get_world_size()
+
+    q_proj = nn.Linear(hidden_dim, hidden_dim)
+    k_proj = nn.Linear(hidden_dim, hidden_dim)
+    v_proj = nn.Linear(hidden_dim, hidden_dim)
+    out_proj = nn.Linear(hidden_dim, hidden_dim)
```

Comment:
No need to create projection layers. You can generation hidden states directly

Reply:
having rewritten the test script